### PR TITLE
New analysis

### PR DESCRIPTION
**Pull request by ldamore**
## Description
This PR introduces two additional benchmark analysis proposed by B. Swope and X. Lucas.

## Installation of `new_analysis` command group in a new conda environment

First, follow the installation procedures for the `openff-benchmark-optimization` environment described in [Deployment Procedure document](https://openforcefield.atlassian.net/wiki/spaces/PS/pages/873922575/Deployment+Procedure#Installation)

Once this is done, you can clone the environment into a new conda environment:

```
conda create -n openff-analysis --clone openff-benchmark-optimization
conda activate openff-analysis
```

Install the `new_analysis` branch from github:

```
git clone https://github.com/openforcefield/openff-benchmark
cd openff-benchmark
git checkout --track origin/new-analysis
pip install -e .
```

## General comments
The two new analysis are typically executed at *point (5)* of the [Optimization Benchmark Protocol](https://openforcefield.atlassian.net/wiki/spaces/PS/pages/971898891/Optimization+Benchmarking+Protocol+-+Season+1)

With `openff-benchmark report swope` the analysis proposed by B. Swope is executed. The command accepts the paths of the optimized molecules obtained from the optimization step. Additionally, the reference method (b3lyp-d3bj by default) and an output directory can be specified.

  `openff-benchmark report swope --input-path 4-compute-qm-filtered --input-path 4-compute-mm-filtered --ref-method b3lyp-d3bj  --output-directory 5-results-swope`

The command creates one output csv file per method, which are named `swope_<method>.csv`, i.e. `swope_openff-1.0.0.csv`

The analysis proposed by B. Swope operates as follows:

- For each molecule MOL-00001-XX.sdf

  - Report the relative energy (**dE**) of each MM optimized conformer (**MM_conf**) with respect to the MM optimized conformer which is the global minimum (**MM_min**)

  - **dE = E<sub>MM_conf</sub> - E<sub>MM_min</sub>**

  - Report the RMSD between each **MM_conf** and the QM optimized conformer which is the global minimum (**QM_min**)

  - **RMSD (MM_conf | QM_min)**

With `openff-benchmark report lucas` the analysis proposed by X. Lucas is executed. The command accepts the paths of the optimized molecules obtained from the optimization step. Additionally, the reference method (b3lyp-d3bj by default) and an output directory can be specified.

`openff-benchmark report lucas --input-path 4-compute-qm-filtered --input-path 4-compute-mm-filtered --ref-method b3lyp-d3bj --output-directory 5-results-lucas`

The command creates one output csv file per method, which are named `lucas_<method>.csv`, i.e. `lucas_openff-1.0.0.csv`

The analysis proposed by X. Lucas operates as follows:

- For each molecule MOL-00001-XX.sdf

  - Find the MM reference conformer (**MM_ref**) with the lowest RMSD with respect to **QM_min**

  - Report the relative energy (**dE**) and RMSD between **MM_ref** and **MM_min**

  - **dE = E<sub>MM_ref</sub> - E<sub>MM_min</sub>**

  - **RMSD (MM_ref | MM_min)**

The final `openff-benchmark report plots-swope` and `openff-benchmark report plots-lucas` commands take the directories containing the csv files as an input (output of `5-results-swope` or `5-results-lucas`).

For `5-results-swope` an `rmsd-cutoff` and `de-cutoff` should be set. 

`openff-benchmark report plots-swope --input-path 5-results-swope/ --de-cutoff 1.5 --rmsd-cutoff 1.0 `

The algorithm will generate a ridge plot of all the conformers within the rmsd-cutoff for a range of dE values, and another ridge plot of all the conformers within the de-cutoff for a range of rsmd values.

For `5-results-lucas`

`openff-benchmark report plots-lucas --input-path 5-results-lucas/`

The algorithm will generate similar plots as for `compare-forcefields` and `match_minima`

### Please note
Likewise `openff-benchmark report match-minima`, also `openff-benchmark report lucas` matches the conformers by rmsd and this step is quite time consuming. However, the intersection method added in this PR now allows the user to run the analysis on each different FF method as separate task, speeding up all the process. e.g.

```
for mm_path in `ls -d 4-compute-mm-filtered/*`; do 
   openff-benchmark report match-minima --input-path 4-compute-qm-filtered \
                                        --input-path $mm_path \
                                        --output-directory 5-results-lucas &  done
```

In addition, the QM-to-QM comparison will be skipped.
Please note that now the plot commands runs _without_ specifying the reference method, e.g.

```
openff-benchmark report plots-lucas --input-path 5-results-lucas/
```

## Todos

  - [x] Fix an error that prevent the drawing of density plots when using the intersection of methods at the plotting stage.
  - [x] Introduced the intersection of methods at the plotting stage. 
  - [ ] Increase the size of y_ticks_labels to the same as x_ticks_labels.

## Questions
- [ ] Question1

## Status
- [ x] Ready to go

### CODE REVIEW

**Comment by dotsdl** [comment]
Awesome, thanks @ldamore! Reviewing!

File: openff/benchmark/analysis/new_draw.py [hunk 1]
```diff
@@ -0,0 +1,973 @@
...
+    plt.xticks(fontsize=ridgedict["xfontsize"])
+    #plt.yticks(fontsize=ridgedict["xfontsize"])
+
+    # save with transparency for overlapping plots
+    plt.savefig(out_file, transparent=True, bbox_inches="tight")
+    plt.clf()
+
+
+def plot_lucas(results_dir, output_directory):
+    global results
```
Comment (dotsdl) [inline openff/benchmark/analysis/new_draw.py:273]: Remove use of `global`. Shouldn't be necessary and we really don't want global state on the module.

File: openff/benchmark/analysis/new_draw.py [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+
+#            draw_corr(
+#                result.loc[:, 'tfd'],
+#                results[method2].loc[:, 'tfd'],
+#                f"{method2} vs. {method}",
+#                f"TFD {method}",
+#                f"TFD {method2}",
+#                os.path.join(output_directory, f"fig_scatter_tfd_{method2}_{method}.png"),
+#                "paper"
+#            )
```
Comment (dotsdl) [inline openff/benchmark/analysis/new_draw.py:349]: Remove if not needed.

File: openff/benchmark/analysis/new_draw.py [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+            result.loc[:, 'rmsd (mm_ref / mm_min)'],
+            result.loc[:,'dE (mm_ref - mm_min)'],
+            method,
+            'rmsd ($\AA$)',
+            'dE (kcal/mol)',
+            os.path.join(output_directory, f'fig_{method}_density_rmsd_dde_linear.png'),
+            what_for="talk",
+            # x_range=(0,3.7),
+            # y_range=(-30,30),
+            # z_range=(-260,5200),
```
Comment (dotsdl) [inline openff/benchmark/analysis/new_draw.py:324]: Remove if not needed.

File: openff/benchmark/analysis/new_draw.py [hunk 4]
```diff
@@ -0,0 +1,973 @@
...
+    colors = sns.color_palette(my_cmap)
+
+    #temp = []
+    #for method, result in dataframes.items():
+    #    result = result.rename(columns={key: x_label})
+    #    result["method"] = method
+    #    temp.append(result)
+        
+        # list of dataframes concatenated to single dataframe
+    #df = pd.concat(temp, ignore_index=True)
```
Comment (dotsdl) [inline openff/benchmark/analysis/new_draw.py:152]: Remove if not needed.

File: openff/benchmark/analysis/new_draw.py [hunk 5]
```diff
@@ -0,0 +1,973 @@
...
+    key,
+    x_label,
+    out_file,
+    what_for="paper",
+    bw="hist",
+    same_subplot=False,
+    sym_log=False,
+    hist_range=(-15, 15)
+):
+    # Define and use a simple function to label the plot in axes coordinates
```
Comment (dotsdl) [inline openff/benchmark/analysis/new_draw.py:100]: Can you add a docstring indicating how this function differs from `draw.draw_ridge`?

File: openff/benchmark/analysis/new_draw.py [hunk 6]
```diff
@@ -0,0 +1,973 @@
...
+        g.fig.subplots_adjust(hspace=-1.0)
+    
+    # Remove axes details that don't play well with overlap
+    g.set_titles("")
+    #    g.set(yticks=[])
+    # g.despine(bottom=True, left=True)
+    # ax = plt.gca()
+    # ax.spines['left'].set_visible(True)
+    # ax.spines['left'].set_position('zero')
+    # ax.set_yticks([0.4])
```
Comment (dotsdl) [inline openff/benchmark/analysis/new_draw.py:255]: Remove if not needed.

File: openff/benchmark/analysis/new_analysis.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+import pandas as pd
+import warnings
+
+from rdkit import Chem
+from rdkit.Chem import rdMolAlign
+from tqdm import tqdm
+
+from . import metrics, readwrite
+
+def lucas(input_path, ref_method, output_directory="./5-results-lucas"):
```
Comment (dotsdl) [inline openff/benchmark/analysis/new_analysis.py:26]: Can you add a docstring to this function describing what it takes, what it does, etc.? You can use your notes from the PR and sculpt them into shape for here.

Same for the `swope` function.

File: openff/benchmark/cli.py [hunk 8]
```diff
@@ -777,6 +777,41 @@ def plots(input_path, output_directory):
 
     draw.plot_compare_ffs(input_path, output_directory)
 
+@report.command()
```
Comment (dotsdl) [inline openff/benchmark/cli.py:780]: Can you add docstrings for each of these four commands? See the other commands in the CLI for inspiration on how to structure this.

**Review by dotsdl — CHANGES_REQUESTED** [review]
Thank you @ldamore for putting this together! I have added inline comments that should be addressed before we proceed to merge. Happy to assist with any of these live as well!

**Comment by ldamore** [comment]
Closed this PR since [PR#95](https://github.com/openforcefield/openff-benchmark/pull/95) resolved the points above.

**Pull request closed without merging**